homeassistant/components/valve/__init__.py

```diff
@@ -0,0 +1,281 @@
+        raise ValueError(f"'reports_position' not set for {self.entity_id}.")
+
+    @property
+    def current_valve_position(self) -> int | None:
+        """Return current position of valve.
+
+        None is unknown, 0 is closed, 100 is fully open.
+        """
+        return self._attr_current_valve_position
+
+    @property
+    def device_class(self) -> ValveDeviceClass | None:
+        """Return the class of this entity."""
+        if hasattr(self, "_attr_device_class"):
+            return self._attr_device_class
+        if hasattr(self, "entity_description"):
+            return self.entity_description.device_class
+        return None
+
+    @property
+    @final
+    def state(self) -> str | None:
+        """Return the state of the valve."""
+        reports_position = self.reports_position
+        if self.is_opening:
+            self._valve_is_last_toggle_direction_open = True
+            return STATE_OPENING
+        if self.is_closing:
+            self._valve_is_last_toggle_direction_open = False
+            return STATE_CLOSING
+        if reports_position is True:
+            if (current_valve_position := self.current_valve_position) is None:
+                return None
+            closed: bool | None = current_valve_position == 0
+            return STATE_CLOSED if closed else STATE_OPEN
+        if (closed := self.is_closed) is None:
+            return None
+        return STATE_CLOSED if self.is_closed else STATE_OPEN
+
+    @final
+    @property
+    def state_attributes(self) -> dict[str, Any]:
+        """Return the state attributes."""
+
+        return {ATTR_CURRENT_POSITION: self.current_valve_position}
+
+    @property
+    def supported_features(self) -> ValveEntityFeature:
+        """Flag supported features."""
+        return self._attr_supported_features
+
+    @property
+    def is_opening(self) -> bool | None:
+        """Return if the valve is opening or not."""
+        return self._attr_is_opening
+
+    @property
+    def is_closing(self) -> bool | None:
+        """Return if the valve is closing or not."""
+        return self._attr_is_closing
+
+    @property
+    def is_closed(self) -> bool | None:
+        """Return if the valve is closed or not."""
+        if self.reports_position:
+            if self.current_valve_position is None:
+                return None
+            return self.current_valve_position == 0
+        return self._attr_is_closed
+
+    def open_valve(self) -> None:
+        """Open the valve."""
+        if not ValveEntityFeature.SET_POSITION | self.supported_features:
+            raise NotImplementedError()
+        self.set_valve_position(100)
+
+    async def async_open_valve(self) -> None:
+        """Open the valve."""
+        await self.hass.async_add_executor_job(ft.partial(self.open_valve))
+
+    def close_valve(self) -> None:
+        """Close valve."""
+        if ValveEntityFeature.SET_POSITION | self.supported_features:
```

Comment:
Same comment as for `open_valve`.

Reply:
The comment for `open_valve` was that we can invert the check and outdent below.

https://github.com/home-assistant/core/pull/102184#discussion_r1403770012